src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -456,10 +532,27 @@ private static bool IsSolutionOpenFromVSSolution(IVsSolution vsSolution)
 
         public async Task<string> GetSolutionFilePathAsync()
         {
-            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-            var vsSolution = await _asyncVSSolution.GetValueAsync();
-            return (string)GetVSSolutionProperty(vsSolution, (int)__VSPROPID.VSPROPID_SolutionFileName);
-
+            if (_solutionEventsCookie != 0)
+            {
+                return _solutionFileName;
+            }
+            else
+            {
+                if (NuGetUIThreadHelper.JoinableTaskFactory.Context.IsOnMainThread)
```

Comment:
This is checking if we're on the main thread and then switching to the main thread?  Or this is checking if the current thread can switch to the main thread?

Reply:
Neither. `IVsBackgroundSolution` thows when called on the UI thread. Not a design decision I would have made, but it is what it is. Therefore `JoinableTaskContext.IsOnMainThread` checks if the current execution thread is the main thread, so we can choose between using `IVsSolution` (UI thread only) or `IVsBackgroundSolution` (non-UI thread only). The goal is to avoid switching threads at all, to stay on the current thread.

The `jtf.SwitchToMainThreadAsync`()` on the line below is not necessary, except the VS Threading Analyzers warn (promoted to error) if we don't. I'm going to give feedback about this. My understanding is that it's supposed to be "fast" to use `SwitchToUIThreadAsync` when you're already on the UI thread, but maybe suppressing the warning is better?  Looking at the implementation of `SwitchToUIThreadAsync` it uses a `alwaysYield: true` parameter, which makes me think it will yield the thread, then wait to be rescheduled before completing work, which is unnecessary latency.

Reply:
Got it.  You are suppressing a warning in the `else`, as you said, maybe it would be better to suppress the warning here as well?  I don't feel strongly either way, this just stood out to me but now I understand.

Reply:
Currently `IVsBackgroundSolution` is in the Microsoft.Internal namespace, and talking to the author, it's going to be moved to be more obviously useful for others to use. I'm worried about it being a breaking change, so I'm removing all usage of it for now. Therefore, we won't be using `JoinableTasdkContext.IsOnMainThread` any longer.